### PR TITLE
#0: Adding flush between mcast write and mcast sem set

### PR DESCRIPTION
**Pull request by abhullar-tt**
Missed flush as part of https://github.com/tenstorrent/tt-metal/pull/12627

### Checklist
- [x] [Post commit CI passes](https://github.com/tenstorrent/tt-metal/actions/runs/10859620098) (updated https://github.com/tenstorrent/tt-metal/actions/runs/11392537298)
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/10859622191)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/matmul/device/kernels/dataflow/reader_bmm_tile_layout_in1_sender_writer_padding.cpp [hunk 1]
```diff
@@ -272,8 +272,12 @@ void kernel_main() {
...
                 in1_start_address, in1_multicast_data_addr, in1_block_size_bytes, in1_mcast_num_cores, true, true);
 
-            // Note: no need for write barrier, since these two multicasts are done on the same noc id, same vc, same
-            // cmd_buf Also, this only works because we are setting VCs statically (using NOC_CMD_STATIC_VC).
+            // Note: no need for write barrier, since these two multicasts are done on the same noc id and same vc even though cmd bufs are different
+            // Also, this only works because we are setting VCs statically (using NOC_CMD_STATIC_VC).
+#ifdef ARCH_BLACKHOLE
+            // On Blackhole the flush is needed because the commands go into separate cmd buffer FIFOs and may not be sent in order they are issued
+            noc_async_writes_flushed();
+#endif
```
Comment (yugaoTT) [inline ttnn/cpp/ttnn/operations/matmul/device/kernels/dataflow/reader_bmm_tile_layout_in1_sender_writer_padding.cpp:280]: should this flush after semaphore call ?

**Review by yugaoTT — CHANGES_REQUESTED** [review]
I wonder if change to use same cmd_buf for mcast data and mcast sem can solve this or not? 
If not, I wonder if it's a bigger problem with bmd_buf_fifo and potentially could affect all noc reads/writes

**Comment by bbradelTT** [comment]
What is the criteria under which this change is needed?

I think there are more noc_semaphore_set_multicast calls in matmul kernels.

```
tt-metal % cd ttnn/cpp/ttnn/operations/matmul/device 
bbradel@Borys-Bradel's-Macbook-Pro device % grep noc_semaphore_set_multicast -r *
kernels/dataflow/reader_bmm_tile_layout_in0_sender_dram_sharded.cpp:            noc_semaphore_set_multicast(
kernels/dataflow/reader_bmm_tile_layout_in0_sender_dram_sharded.cpp:                noc_semaphore_set_multicast_loopback_src(
kernels/dataflow/reader_bmm_tile_layout_in1_sender_writer_padding.cpp:            noc_semaphore_set_multicast(
kernels/dataflow/reader_bmm_tile_layout_in1_sender_writer_padding.cpp:            noc_semaphore_set_multicast(
kernels/dataflow/reader_bmm_tile_layout_in0_sender_padding.cpp:            noc_semaphore_set_multicast(
kernels/dataflow/reader_bmm_tile_layout_in0_sender_padding_block_sharded.cpp:            noc_semaphore_set_multicast(
kernels/dataflow/reader_bmm_tile_layout_in0_sender_receiver_padding_block_sharded.cpp:                        // noc_semaphore_set_multicast_loopback_src is a no-op in this case.
kernels/dataflow/reader_bmm_tile_layout_in0_sender_receiver_padding_block_sharded.cpp:                        noc_semaphore_set_multicast_loopback_src(
kernels/dataflow/reader_bmm_tile_layout_in0_sender_receiver_padding_block_sharded.cpp:                    noc_semaphore_set_multicast(
```

**Comment by abhullar-tt** [comment]
> What is the criteria under which this change is needed?
> 
> I think there are more noc_semaphore_set_multicast calls in matmul kernels.
> 
> ```
> tt-metal % cd ttnn/cpp/ttnn/operations/matmul/device 
> bbradel@Borys-Bradel's-Macbook-Pro device % grep noc_semaphore_set_multicast -r *
> kernels/dataflow/reader_bmm_tile_layout_in0_sender_dram_sharded.cpp:            noc_semaphore_set_multicast(
> kernels/dataflow/reader_bmm_tile_layout_in0_sender_dram_sharded.cpp:                noc_semaphore_set_multicast_loopback_src(
> kernels/dataflow/reader_bmm_tile_layout_in1_sender_writer_padding.cpp:            noc_semaphore_set_multicast(
> kernels/dataflow/reader_bmm_tile_layout_in1_sender_writer_padding.cpp:            noc_semaphore_set_multicast(
> kernels/dataflow/reader_bmm_tile_layout_in0_sender_padding.cpp:            noc_semaphore_set_multicast(
> kernels/dataflow/reader_bmm_tile_layout_in0_sender_padding_block_sharded.cpp:            noc_semaphore_set_multicast(
> kernels/dataflow/reader_bmm_tile_layout_in0_sender_receiver_padding_block_sharded.cpp:                        // noc_semaphore_set_multicast_loopback_src is a no-op in this case.
> kernels/dataflow/reader_bmm_tile_layout_in0_sender_receiver_padding_block_sharded.cpp:                        noc_semaphore_set_multicast_loopback_src(
> kernels/dataflow/reader_bmm_tile_layout_in0_sender_receiver_padding_block_sharded.cpp:                    noc_semaphore_set_multicast(
> ```

The issue seems to be occurring when we have a mcast write followed by mcast sem set without a flush/barrier in between. I can take a look at updating these cases as well. This is just a temp workaround until I figure out why this is needed.

File: ttnn/cpp/ttnn/operations/matmul/device/kernels/dataflow/reader_bmm_tile_layout_in1_sender_writer_padding.cpp (hunk 1 above)
Comment (abhullar-tt) [reply ttnn/cpp/ttnn/operations/matmul/device/kernels/dataflow/reader_bmm_tile_layout_in1_sender_writer_padding.cpp:280]: it seems the flush is needed before the sem call

**Comment by abhullar-tt** [comment]
> I wonder if change to use same cmd_buf for mcast data and mcast sem can solve this or not? If not, I wonder if it's a bigger problem with bmd_buf_fifo and potentially could affect all noc reads/writes

I experimented with this and it seems like using the same cmd buffer also fixes this issue

**Comment by yugaoTT** [comment]
@abhullar-tt I think fix with using same cmd buf makes more sense, and also not losing any perf since the flush can cause the receiver cores to wait longer time. Could you please remove the flush and use the same cmd buf fix instead?

**Review by bbradelTT — APPROVED** [review]

**Comment by bbradelTT** [comment]
I had to add more flushes. See https://github.com/tenstorrent/tt-metal/issues/12600#issuecomment-2420401891 for details.

**Review by yugaoTT — APPROVED** [review]
could you also update the other files with mcast + mcast semaphore?

**Comment by abhullar-tt** [comment]
> could you also update the other files with mcast + mcast semaphore?

@bbradelTT is following up with another PR to address those

**Pull request merged**